### ISSUE TITLE
Exec semantics

### ISSUE THREAD

**Issue by vishh** [opencontainers/runtime-spec#345]
The current Spec attempts to express `exec` functionality by splitting the container into a sandbox and a process. This split fails to address some of the use cases like as follows - 
1. Imposing additional resource restrictions on the exec process.
2. Imposing alternate/additional security policy on the exec process.

It is proving to be difficult to represent all the use cases and also preserve the semantics around the sandbox and the process.
Have we considered representing `exec` as a separate container with a process? That separate container can share some parts of the existing container's sandbox.
IIRC, the Spec tried to represent `exec` as a separate container in the beginning, but I cannot recall when it diverged from the original plan.

cc @crosbymichael @mrunalp

**Comment by wking** [comment]
On Wed, Mar 16, 2016 at 12:37:22PM -0700, Vish Kannan wrote:
> IIRC, the Spec tried to represent `exec` as a separate container in
> the beginning, but I cannot recall when it diverged from the
> original plan.

‘exec’ only landed three weeks ago with #225, so there's not a lot of
history in the spec itself.  There has been wording that looks sort of
like separate create / exec since 77d44b10 (Update runtime.md,
2015-06-16) [1], but that's from the ancient past before changes were
publically motivated and the wording went away with the initial
lifecycle docs in #231.

Public out-of-spec discussion around separating sandbox creation from
process execution goes back a lot further (e.g. see [2,3,4,5]).  I
think the main point in favor of distinguishing between sandboxes
(create) and processes (exec) is that it makes it easy to launch new
processes in an existing container [6].

> Have we considered representing `exec` as a separate container with
> a process?

I think the namespace creation and cgroup joining in the current spec
should allow you to:

1. Join an existing container (or not).
2. Create a new container (or not).
3. Execute a new process inside (1) and/or (2).

As far as cgroups are concerned, join-and-create-a-subgroup is easy
with cgroupsPath (just use ‘{existing-parent-path}/{new-child-path}’,
see #68).  Namespaces are a bit trickier, because of the very-narrow
visibility scoping in the state registry [7,8].  It's possible that:

  $ nsenter <parent-namespaces> funC start <container-id> .

would create a new container whose state was only accessible via:

  $ nsenter <parent-namespaces> funC state <container-id>

and not directly via:

  $ funC state <container-id>

An extra nsenter call doesn't seem like that big a deal though.

And this sort of nested-namespace approach will also make clarifying
PID namespaces more important [9].

[1]: https://github.com/opencontainers/specs/commit/77d44b10d5df53ee63f0768cd0a29ef49bad56b6#diff-b84a8d65d8ed53f4794cd2db7e8ea731R478
[2]: https://groups.google.com/a/opencontainers.org/d/msg/dev/BIxya5eSNLo/Jz5yfRVpFAAJ
     Subject: Re: Specifying the runtime's command-line interface
     Date: Wed, 26 Aug 2015 00:01:48 -0700 (PDT)
     Message-Id: <d7dd9194-8368-4ea0-a0fb-1a9b4fb139b1@opencontainers.org>
[3]: https://groups.google.com/a/opencontainers.org/forum/#!topic/dev/kPWu-ZavQjs
     Date: Wed, 26 Aug 2015 13:40:47 -0700
     Subject: Splitting container.json from application.json
     Message-ID: <20150826204047.GY21585@odin.tremily.us>
[4]: https://groups.google.com/a/opencontainers.org/forum/#!topic/dev/frUXLljXy8Y
     Subject: Documenting container lifecycles
     Date: Mon, 14 Sep 2015 20:47:17 -0700
     Message-ID: <20150915034717.GO18018@odin.tremily.us>
[5]: https://github.com/opencontainers/specs/issues/299
[6]: https://github.com/opencontainers/specs/wiki/MeetingMinutes:-2015-09-16#minutes
[7]: https://github.com/opencontainers/specs/pull/225/#issuecomment-172633075
[8]: https://github.com/opencontainers/specs/pull/225/files#diff-b84a8d65d8ed53f4794cd2db7e8ea731R3
[9]: https://groups.google.com/a/opencontainers.org/forum/#!topic/dev/ujtABQoCmgk
     Subject: Linux: Adding the PID namespace inode to state JSON
     Date: Wed, 30 Dec 2015 21:34:57 -0800
     Message-ID: <20151231053457.GG3680@odin.tremily.us>

**Comment by vishh** [comment]
Thanks for the detailed response @wking!
To rephrase, I guess my proposal is to not include [`exec`](https://github.com/opencontainers/specs/blob/fed01f4d97e0b58ad573801d74aa6485754117d5/runtime.md#exec) as a first class entity in the Spec and instead define a standard way for implementing it using the Spec.

**Comment by crosbymichael** [comment]
@vishh i have always said i don't think exec should be in the spec because ppl want too many different things for it.  This would be a much better solution instead of trying to code in every use case into this one feature.

**Comment by vishh** [comment]
@crosbymichael: Yeah. Agreed. It's one thing to come up with a suggestion for implementing exec, but making it a standardized feature doesn't make sense.

ping @mrunalp @duglin @julz for input here.

**Comment by crosbymichael** [comment]
@vishh same.  just because we choose not to standardize on exec because of the variations does not mean runtimes cannot implement it, it means they have more freedom to implement it to their specific requirements.

**Comment by wking** [comment]
On Wed, Mar 16, 2016 at 04:02:51PM -0700, Vish Kannan wrote:
> To rephrase, I guess my proposal is to not include `exec` as a first
> class entity in the Spec and instead define a standard way for
> implementing it using the Spec.

This is fine with me, but see also [1,2,3].  ‘exec’ didn't get much
airtime in #225 (I don't see a reply to [4]).  Whichever way we go, it
would be nice if changes to operations landed with clear motivation in
the associated commit and/or PR, so we don't remove operations within
a month of adding them as frequently ;).

[1]: https://github.com/opencontainers/specs/issues/299#issuecomment-174032625
[2]: http://ircbot.wl.linuxfoundation.org/eavesdrop/%23opencontainers/%23opencontainers.2016-03-09.log.html#t2016-03-09T18:26:49
[3]: http://ircbot.wl.linuxfoundation.org/eavesdrop/%23opencontainers/%23opencontainers.2016-03-09.log.html#t2016-03-09T18:14:40
[4]: https://github.com/opencontainers/specs/pull/225#issuecomment-172563011

**Comment by mrunalp** [comment]
@vishh I agree with @crosbymichael. I think there are too many variations and it might make sense to leave it up to the runtimes to implement it as required.

**Comment by julz** [comment]
yeah I'm slowly changing my mind on this :). 

I was initially in favour of standardising exec because it makes certain common things super easy (and as a higher-level system that wants the freedom to easily swap out runtimes, I really only want to rely on standardised bits). But I can't think of a case where there's something I can achieve with `exec` that I can't achieve by creating a second container that shares all the namespaces/cgroups/etc of the main container -- and with more power especially in the case of cgroups. So maybe a high level system that wants to support the maximum number of backends can/should just rely on `create` and ignore `exec` entirely. And @crosbymichael / @vishh are clearly right that the further we push on `exec` the closer we get to a bad version of just doing that in the first place :). 

So +1 from me, as long as we agree (I think everyone does anyway?) that anything possible with exec should always also be possible via create/start with an appropriate config.

**Comment by vishh** [comment]
@julz @mrunalp: Appreciate the feedback. If @duglin also doesn't have any objections to this change, I can post a patch to make exec not part of the UX and add documentation explaining how it can be achieved using the Spec.

**Comment by duglin** [comment]
@vishh as @julz said, as long as we can get the same net result with a secondary create/start then I'm ok with it. After reading your proposed doc changes for how all of this will be expressed in the config.json I'd like to explore some runc command line niceties to make it easier for people - or at least try, but that's secondary.

**Comment by mrunalp** [comment]
@duglin The way I see it, I think runc can continue doing what it is doing today and then additional features as necessary.

**Comment by duglin** [comment]
@mrunalp do you mean expose it as "runc exec .." but under the covers do some kind of "join" to the existing namespaces/etc... ?

**Comment by mrunalp** [comment]
@duglin Yes.

**Comment by duglin** [comment]
@mrunalp that would be excellent!  Does this mean we're closer to adopting the create/start split?  :-)

**Comment by mrunalp** [comment]
@duglin Still needs to be reviewed :D However, exec can be achieved either ways :)

**Comment by julz** [comment]
Yeah to be clear I think runC exec should stay, just I agree we don't need to specify it. Imo instead we should doc (in the spec) how to specify a config to exec a process in an existing container. Also seems potentially useful either in ocitools or runC to have a command to generate an exec config for a running container (but this isn't super hard to do manually anyway).

**Comment by wking** [comment]
After #388, is there anything left to do here?

**Comment by hqhq** [comment]
@wking It should be safe to close now.
